src/System.Management.Automation/engine/parser/DynamicKeyword.cs

```diff
@@ -0,0 +1,2759 @@
+                    case DynamicKeywordUseMode.Required:
+                    case DynamicKeywordUseMode.RequiredMany:
+                        if (!ScopeSeenDynamicKeywords.Peek().Contains(keyword))
+                        {
+                            unusedKeywords.Add(keyword);
+                        }
+                        break;
+                }
+            }
+
+            return unusedKeywords;
+        }
+
+        /// <summary>
+        /// Wipe this namespace clean. If this is tried while in a scope, we will be in an
+        /// invalid state, hence it is disallowed
+        /// </summary>
+        public void Reset()
+        {
+            if (DynamicKeywordScope.Count > 0)
+            {
+                throw new PSInvalidOperationException("Cannot reset the DynamicKeyword namespace while in a DynamicKeyword scope");
+            }
+
+            _dynamicKeywordScope = new Stack<DynamicKeyword>();
+            _scopeSeenDynamicKeywords = new Stack<HashSet<DynamicKeyword>>(new[] { new HashSet<DynamicKeyword>() });
+            _globalKeywords = new Dictionary<string, DynamicKeyword>();
+        }
+    }
+
+    #endregion /* Dynamic Keyword Namespacing */
+
+    /// <summary>
+    /// Defines the name modes for a dynamic keyword. A name expression may be required, optional or not permitted.
+    /// </summary>
+    public enum DynamicKeywordNameMode
+    {
+        /// <summary>
+        /// This keyword does not take a name value
+        /// </summary>
+        NoName = 0,
+        /// <summary>
+        /// Name must be present and simple non-empty bare word
+        /// </summary>
+        SimpleNameRequired = 1,
+        /// <summary>
+        /// Name must be present but can also be an expression
+        /// </summary>
+        NameRequired = 2,
+        /// <summary>
+        /// Name may be optionally present, but if it is present, it must be a non-empty bare word.
+        /// </summary>
+        SimpleOptionalName = 3,
+        /// <summary>
+        /// Name may be optionally present, expression or bare word
+        /// </summary>
+        OptionalName = 4,
+    };
+
+    /// <summary>
+    /// Defines the body mode for a dynamic keyword. It can be a scriptblock, hashtable or command which means no body
+    /// </summary>
+    public enum DynamicKeywordBodyMode
+    {
+        /// <summary>
+        /// The keyword act like a command
+        /// </summary>
+        Command = 0,
+        /// <summary>
+        /// The keyword has a scriptblock body
+        /// </summary>
+        ScriptBlock = 1,
+        /// <summary>
+        /// The keyword has hashtable body
+        /// </summary>
+        Hashtable = 2,
+    }
+
+    /// <summary>
+    /// Defines the use semantics of a dynamic keyword for a given block
+    /// </summary>
+    public enum DynamicKeywordUseMode
+    {
+        /// <summary>
+        /// The keyword must be used exactly once in a block
+        /// </summary>
+        Required = 0,
```

Comment:
Yeah, to be honest, I'm not entirely sure about the `UseMode` concept as a whole. I introduced it because I could conceive of cases where declarations only make sense once (such as a prelude statement to do something like set up common instance infrastructure or initialize a format variable), or where a group of statements need to have a keyword present to make sense. I only introduced the `Many` cases after my restricted definition of the singular ones.

I think there are three choices:
- Only have `Required` and `Optional` with the meanings you give
- Change to a number range to specify how many times a keyword can be used (with a value like `-1` or `Int32.MinValue` meaning unlimited)
- Get rid of `UseMode`

I think the second option is probably fraught with corner cases and may not have good enough use cases to justify. But I suggest it because the spirit of a DSL to me seems to be to want to make it easy to write valid declarations and hard to write invalid ones -- so granularity in `UseMode` allows the encoding of stronger semantic checks.